// react-easy-chart/lib/line-chart/hybrid/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _common = require('./LineChart/common');

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _radium = require('radium');

var _lodash = require('lodash.merge');

var _lodash2 = _interopRequireDefault(_lodash);

var _LineChart = require('./LineChart');

var _LineChart2 = _interopRequireDefault(_LineChart);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var LineChart = function (_React$Component) {
  (0, _inherits3.default)(LineChart, _React$Component);

  function LineChart(props) {
    (0, _classCallCheck3.default)(this, LineChart);

    var _this = (0, _possibleConstructorReturn3.default)(this, (LineChart.__proto__ || (0, _getPrototypeOf2.default)(LineChart)).call(this, props));

    _this.uid = (0, _common.createUniqueID)();
    return _this;
  }

  (0, _createClass3.default)(LineChart, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      var chart = this.refs.chart;


      var props = this.props;

      _LineChart2.default.initialise(chart, props);
    }
  }, {
    key: 'componentDidUpdate',
    value: function componentDidUpdate() {
      var chart = this.refs.chart;


      var props = this.props;

      _LineChart2.default.transition(chart, props);
    }
  }, {
    key: 'createStyle',
    value: function createStyle() {
      var _props = this.props,
          style = _props.style,
          grid = _props.grid,
          vGrid = _props.vGrid,
          hGrid = _props.hGrid,
          orient = _props.orient;


      var uid = this.uid;
      var scope = '.line-chart-' + uid;
      var defaultAxisStyles = (0, _common.getDefaultAxisStyles)(vGrid || grid, hGrid || grid, orient);
      var rules = (0, _lodash2.default)({}, _common.defaultStyles, style, defaultAxisStyles);

      return _react2.default.createElement(_radium.Style, {
        scopeSelector: scope,
        rules: rules
      });
    }
  }, {
    key: 'render',
    value: function render() {
      var _props2 = this.props,
          width = _props2.width,
          height = _props2.height;


      var uid = this.uid;
      var className = 'line-chart-' + uid;

      return _react2.default.createElement(
        'div',
        { ref: 'lineChart', className: className },
        this.createStyle(),
        _react2.default.createElement('svg', { ref: 'chart', width: width, height: height })
      );
    }
  }]);
  return LineChart;
}(_react2.default.Component);

exports.default = LineChart;


LineChart.propTypes = {
  data: _propTypes2.default.array.isRequired,
  axes: _propTypes2.default.bool,
  grid: _propTypes2.default.bool,
  hGrid: _propTypes2.default.bool,
  vGrid: _propTypes2.default.bool,
  type: _propTypes2.default.string,
  xType: _propTypes2.default.string,
  yType: _propTypes2.default.string,
  interpolate: _propTypes2.default.string,
  margin: _propTypes2.default.shape({
    top: _propTypes2.default.number,
    right: _propTypes2.default.number,
    bottom: _propTypes2.default.number,
    left: _propTypes2.default.number
  }),
  width: _propTypes2.default.number,
  height: _propTypes2.default.number,
  orient: _propTypes2.default.string,
  colors: _propTypes2.default.array,
  strokeWidth: _propTypes2.default.number,
  stroke: _propTypes2.default.object,
  style: _propTypes2.default.object,
  xLabel: _propTypes2.default.string,
  yLabel: _propTypes2.default.string,
  xDomain: _propTypes2.default.array,
  yDomain: _propTypes2.default.array,
  pattern: _propTypes2.default.string,
  xPattern: _propTypes2.default.string,
  yPattern: _propTypes2.default.string
};

LineChart.defaultProps = {
  type: 'linear',
  xType: 'linear',
  yType: 'linear',
  interpolate: 'linear',
  margin: {
    top: 0,
    right: 0,
    bottom: 0,
    left: 0
  },
  orient: 'left',
  colors: [],
  strokeWidth: 2,
  style: {}
};
module.exports = exports['default'];
